File: server_python/amq/list/handlers.py
from amq import sio, db, logger, clientUpdater
from amq.decorators import exceptionHandler

def listPickerHandlers():

    @sio.on('LOGIN_LIST')
    @exceptionHandler
    def login(sid):
        clientUpdater.updateUsers(sid)


    @sio.on('SYNC_FULL_LIST')
    @exceptionHandler
    def fullListRequest(sid):
        logger.writeLog('LIST002', { 'id': sid })
        clientUpdater.updateSongList(sid)

    @sio.on('SYNC_USER_LIST')
    @exceptionHandler
    def userListRequest(sid, user):
        logger.writeLog('LIST003', { 'id': sid, 'filename': user })
        clientUpdater.updateUserList(user, sid)

    @sio.on('ADD_SONG')
    @exceptionHandler
    def addSong(sid, song, user):
        anime = song.get('anime', [''])[0]
        title = song.get('title', '')
        songId = song.get('songId', '')
        logger.writeLog('LIST004', { 'id': sid,
                                        'anime': anime,
                                        'title': title,
                                        'songId': songId,
                                        'user': user })
        db.addUserSong(user, songId)
        clientUpdater.updateUserList(user)
        clientUpdater.sendNotification('success', f'{anime}: {title} added', sid)


    @sio.on('REMOVE_SONG')
    @exceptionHandler
    def removeSong(sid, song, user):
        anime = song.get('anime', [''])[0]
        title = song.get('title', '')
        songId = song.get('songId', '')
        logger.writeLog('LIST005', { 'id': sid,
                                    'anime': anime,
                                    'title': title,
                                    'songId': songId,
                                    'filename': user })
        db.removeUserSong(user, songId)
        clientUpdater.updateUserList(user)
        clientUpdater.sendNotification('success', f'{anime}: {title} removed', sid)

File: server_ts/src/list/handlers.ts
import socketio = require('socket.io')
import { AMQLogger } from '../shared/logging/logger'
import { AMQDatabase } from '../database'
import { MessageEmitter } from '../shared/handlers'
import { exceptionHandler } from '../shared/exceptions'
import { Song } from '../shared/interfaces'

class ListPickerHandler {
  _logger: AMQLogger
  _db: AMQDatabase
  _emitter: MessageEmitter

  constructor(logger: AMQLogger, db: AMQDatabase, emitter: MessageEmitter) {
    this._logger = logger
    this._db = db
    this._emitter = emitter
  }

  start(socket: socketio.Socket): void {

    socket.on('LOGIN_LIST', exceptionHandler(socket, () => {
      this._emitter.syncUsers(socket)
    }))

    socket.on('SYNC_FULL_LIST', exceptionHandler(socket, () => {
      this._logger.writeLog('LIST002', { id: socket.id })
      this._emitter.syncSongList(socket)
    }))

    socket.on('ADD_SONG', exceptionHandler(socket, (song: Song, user: string) => {
      let { anime, title, songId } = this._parseSong(song)
      this._logger.writeLog('LIST004', { id: socket.id,
                                         anime: anime,
                                         title: title,
                                         songId: songId,
                                         user: user })
      this._db.addUserSong(user, songId)
      this._emitter.syncUserList(user)
      this._emitter.notification('success', `${anime}: ${title} added`, socket)
    }))

    socket.on('REMOVE_SONG', exceptionHandler(socket, (song: Song, user: string) => {
      let { anime, title, songId } = this._parseSong(song)
      this._logger.writeLog('LIST004', { id: socket.id,
                                         anime: anime,
                                         title: title,
                                         songId: songId,
                                         user: user })
      this._db.removeUserSong(user, songId)
      this._emitter.syncUserList(user)
      this._emitter.notification('success', `${anime}: ${title} removed`, socket)
    }))

  }

  _parseSong(song: Song) {
    let anime = song['anime'][0]
    let title = song['title']
    let songId = song['songId']

    return { anime, title, songId }
  }

}

export { ListPickerHandler }